lib/base.js

```diff
@@ -54,6 +59,75 @@ Base.prototype._checkStatusForError = function(statusCode, body) {
     }
 };
 
+Base.prototype.register = function({username, password, fields}, done) {
+    if (!username) {
+        throw new AirtableError('NOT_FOUND', "Missing parameter 'username' required for Base#register", 404);
+    }
+    if (!password) {
+        throw new AirtableError('NOT_FOUND', "Missing parameter 'password' required for Base#register", 404);
+    }
+    if (this._airtable._endpointUrl.includes('https://api.airtable.com')) {
+        throw new AirtableError('NOT_AUTHORIZED', 'Base#register cannot be used with api.airtable.com. Please configure this Airlock client to use an Airlock endpoint URL.', 403);
+    }
+    if (this._user !== null || this._getLocalStorage('user') !== null) {
+        throw new AirtableError('NOT_AUTHORIZED', 'Already logged in');
+    }
+    runAction(this, 'post', '/__DANGEROUSLY__USE__TABLE__TO__LET__USERS__REGISTER', {}, {username, password, fields}, (err, data) => {
+        // this._user = data;
+        // err = err || this._setLocalStorage('user', data);
+        done(err, data);
+    });
+};
+
+Base.prototype.login = function({username, password}, done) {
+    if (!username) {
+        throw new AirtableError('NOT_FOUND', "Missing parameter 'username' required for Base#login", 404);
+    }
+    if (!password) {
+        throw new AirtableError('NOT_FOUND', "Missing parameter 'password' required for Base#login", 404);
+    }
+    if (this._airtable._endpointUrl.includes('https://api.airtable.com')) {
+        throw new AirtableError('NOT_AUTHORIZED', 'Base#login cannot be used with api.airtable.com. Please configure this Airlock client to use an Airlock endpoint URL.', 403);
+    }
+    if (this._user !== null) {
+        throw new AirtableError('NOT_AUTHORIZED', 'Already logged in');
+    }
+    runAction(this, 'post', '/__DANGEROUSLY__USE__TABLE__TO__LET__USERS__LOGIN', {}, {username, password}, (err, data) => {
+        this._user = {
+            user: {
+                username: data.body.user.username,
+            },
+            token: data.body.token
+        };
+
+        err = err || this._setLocalStorage('user', data);
+        done(err, data);
+    });
+};
+
+
+Base.prototype.logout = function(done) {
+    if (this._airtable._endpointUrl.includes('https://api.airtable.com')) {
+        throw new AirtableError('NOT_AUTHORIZED', 'Base#logout cannot be used with api.airtable.com. Please configure this Airlock client to use an Airlock endpoint URL.', 403);
+    }
+    this._user = null;
+    done(this._setLocalStorage('user', null), {});
+};
+
+Base.prototype._setLocalStorage = function(key, value) {
+    try {
+        localStorage.setItem(key, JSON.stringify(value));
+    } catch (err) {
+        return err;
```

Comment:
This returning error pattern is very Go-like, but might it be better to allow this function to throw? And require functions that use `_setLocalStorage` to handle errors?

Reply:
It's worth noting that `_getLocalStorage` already acts like this too.

Reply:
Good point!